### PR TITLE
Quiz Application

### PR DESCRIPTION
**Pull request by elmq0022**
All the relevant stuff is way on the bottom.   This isn't ready for master yet.

### CODE REVIEW

File: quiz/json_response_mixin.py [hunk 1]
```diff
@@ -0,0 +1,24 @@
+'''
+This module contains a JSON Response Mixin.
+'''
+
+from django.http import JsonResponse
+
+class JSONResponseMixin(object):
```
Comment (jordanbettis) [inline quiz/json_response_mixin.py:7]: Might want to consider putting this in the hanyu app given that you're going to be using it in other packages.

File: quiz/models.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+        quiz_dict = {
+            'user_id': self.user.pk,
+            'uid': str(self.uid),
+            'question': Entry.objects.get(pk=self.correct_answer).simple,
+            'answers': [{'pk':ans.pk, 'definition':ans.entry.definitions} for ans in self.answer_set.all()]
+            }
+        return quiz_dict
+
+
+class Answer(models.Model):
```
Comment (jordanbettis) [inline quiz/models.py]: I'm not entirely sure why both `Answer` and `Quiz` have a field to determine if the answer was correct or not. It seems like that should live in either one or the other.

File: quiz/templates/quiz/quiz.html [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+        <div class="box-footer">
+            <button class="btn btn-primary">Submit</button>
+        </div>
+    </form>
+    </div>
+    <!-- /.box -->
+{% endblock content %}
+
+{% block javascript  %}
+    {{ block.super }}
```
Comment (jordanbettis) [inline quiz/templates/quiz/quiz.html:29]: See I like to do this:
```python
{% block javascript %}
    ...
    {% block extra_javascript %}{% endblock %}
{% endblock %}
```

In the parent template. That way you don't have to override and then call super in the page template, you can just go ahead and override the empty `{% block extra_javascript %}`.

File: quiz/templates/quiz/quiz.html [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+    <script>
+        $(document).ready(main());
+
+        function main(){
+            var url = "/quiz/json";
+
+            $.ajaxSetup({
```
Comment (jordanbettis) [inline quiz/templates/quiz/quiz.html]: You probably want to do the ajaxsetup in the base template so it's available all over your site.

File: quiz/templates/quiz/quiz.html [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+                url,
+            )
+            .done(function(data){
+                setQuestion(data);
+            })
+        }
+
+        function submitAnswer(){
+            var selectedAnswerID = $("input[name='answer']:checked").val();
+            alert(selectedAnswerID);
```
Comment (jordanbettis) [inline quiz/templates/quiz/quiz.html]: I like to use `console.log()` or `console.debug()` for debugging. This puts it into the javascript console.

File: quiz/templates/quiz/quiz.html [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+            return cookieValue;
+        } // end getCookie
+
+        function setQuestion(data){
+            var questionDiv = $("#question");
+            var answerDiv = $("#answers");
+            var answerList = []; 
+            var beg = '<div class="checkbox"><label><input type="radio" name="answer" value=';
+            var mid = '>'
+            var end = '<\/label><\/div>'
```
Comment (jordanbettis) [inline quiz/templates/quiz/quiz.html]: Why not do something like this:
```javascript
var checkboxHtml = function(answer) {
    return ('<div class="checkbox"><label><input type="radio" name="answer" value=' + answer.pk + '>' 
                + answer.definition + '<\/label><\/div>');
};
```

Then on line 106 you can just do:

```javascript
answerList.push(checkboxHtml(data.answers[i]));
```

File: quiz/views.py [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+
+class QuizView(TemplateView):
+    '''
+    Serve the template for quiz application.
+    This is need simply to extend the _base.html template
+    and to include the CSRF token.
+    '''
+    template_name = 'quiz/quiz.html'
+
+    def get_random_quiz(self):
```
Comment (jordanbettis) [inline quiz/views.py]: Is this named correctly?

**Review by jordanbettis — CHANGES_REQUESTED** [review]
This is starting to take shape. You probably want to add "WIP:" to the PR title, to indicate that it shouldn't be merged yet.

File: quiz/models.py (hunk 2 above)
Comment (elmq0022) [reply quiz/models.py]: yes that's redundant.

File: quiz/templates/quiz/quiz.html (hunk 5 above)
Comment (elmq0022) [reply quiz/templates/quiz/quiz.html]: Good to know. Thanks.

File: quiz/json_response_mixin.py (hunk 1 above)
Comment (elmq0022) [reply quiz/json_response_mixin.py:7]: I'm going to put this one on hold.  As I'm ditching the class based view until I get something working.  I'll swing back if there's still time.

File: quiz/templates/quiz/quiz.html (hunk 4 above)
Comment (elmq0022) [reply quiz/templates/quiz/quiz.html]: I moved this long with all the other ajax-y functions.  As you said, there's no reason why they are specific to the quiz template.

File: quiz/templates/quiz/quiz.html (hunk 6 above)
Comment (elmq0022) [reply quiz/templates/quiz/quiz.html]: Done.

File: quiz/views.py (hunk 7 above)
Comment (elmq0022) [reply quiz/views.py]: Scratch work left in the code.  I'll clean this up as I go.

**Comment by elmq0022** [comment]
Jordan, this is still a little rough around the edges, but good enough to merge into *dev* and probably *main* in my opinion.  Please take a look when you can.

**Comment by elmq0022** [comment]
Made final adjustments.  I'm gong to merge this with dev and have it approved pulled from there.  I'm going to start cleaning up a few other things today.

**Pull request merged**